### PR TITLE
handle execve() errors gracefully

### LINKED ISSUE #102: Handle execve() errors gracefully
If simexpal cannot execute an experiment, e.g., because the program is missing, we should still create a `.status` file.

### PR DESCRIPTION
**Pull request by duylethanh**
This PR contains the following:
- fix #102 (the ``.status`` file will be created even if simexpal can not find the executable), this also fixed a bug where experiments were shown as ``started`` when the executable was not found
- added unit test for experiments with missing programs

In the ``launch/common.py`` I mainly wrapped

``` python

child = subprocess.Popen(cmd, cwd=cwd, env=environ,
				stdout=stdout, stderr=stderr)

```

in a ``try/except`` block and put the code below it in the ``else`` case.

### CODE REVIEW

File: simexpal/launch/common.py [hunk 1]
```diff
@@ -424,65 +424,79 @@ def close(self):
...
+		# Log the error traceback.
+		with open(manifest.aux_file_path('stderr'), 'w') as f:
+			f.write(traceback.format_exc())
+
+		# Create the .status file.
+		status_dict = {'timeout': False, 'walltime': 0, 'status': 1, 'signal': None}
+		with open(manifest.output_file_path('status.tmp'), "w") as f:
+			yaml.dump(status_dict, f)
+		os.rename(manifest.output_file_path('status.tmp'), manifest.output_file_path('status'))
 	else:
```
Comment (avdgrinten) [inline simexpal/launch/common.py:442]: I dislike that the entire code is in the `else` block now. Can't we just return in the error case?

File: simexpal/launch/common.py [hunk 2]
```diff
@@ -424,65 +424,79 @@ def close(self):
...
+				stdout=stdout, stderr=stderr)
+	except FileNotFoundError:
+		import traceback
+
+		# Log the error traceback.
+		with open(manifest.aux_file_path('stderr'), 'w') as f:
+			f.write(traceback.format_exc())
+
+		# Create the .status file.
+		status_dict = {'timeout': False, 'walltime': 0, 'status': 1, 'signal': None}
```
Comment (avdgrinten) [inline simexpal/launch/common.py:438]: Not sure if `status`: 1 is a great way to represent this. Maybe we could set `status` = -1 (which should not be a possible return value for successful executions) and add an attribute such as `error: 'executable_not_found'`?

File: simexpal/launch/common.py (hunk 1 above)
Comment (duylethanh) [reply simexpal/launch/common.py:442]: I agree, thats better.

File: simexpal/launch/common.py (hunk 2 above)
Comment (duylethanh) [reply simexpal/launch/common.py:438]: In ``Run._update_status_cache_dict()`` we check for ``status_dict['status'] > 0`` in order to determine failed runs. Also runs that fail without a timeout or signal have ``status_dict['status']== 1``. To use ``status_dict['status']== -1`` we could use the condition ``status_dict['status']> 0 or status_dict['status']< 0`` or alternatively check for the error entry e.g. ``status_dict['status']> 0 or status_dict['error']``. What do you think?

Comment (duylethanh) [reply simexpal/launch/common.py]: I set ``status = -1``  and ``error: 'executable_not_found'`` when the executable is missing and chose the latter approach  i.e. the condition ``status_dict['status']> 0 or status_dict['error']`` to check for this case.

**Comment by duylethanh** [comment]
PR updated.

File: simexpal/base.py [hunk 3]
```diff
@@ -1251,7 +1251,7 @@ def _update_status_cache_dict(self):
 				status = Status.TIMEOUT
 			elif status_dict['signal']:
 				status = Status.KILLED
-			elif status_dict['status'] > 0:
+			elif status_dict['status'] > 0 or status_dict['error']:
```
Comment (avdgrinten) [inline simexpal/base.py:1254]: What if `error` does not exist (e.g., because we're running on a .status file that was generated by an older version)?

Comment (duylethanh) [reply simexpal/base.py]: I changed it to ``status_dict.get('error', None)`` to handle this case.

**Comment by duylethanh** [comment]
PR updated.

**Review by avdgrinten — APPROVED** [review]
LGTM now, thank you.

**Pull request merged**